Allow coin selection with pre-selected inputs

jonathanknowles opened the pull request:
## Issue Number

ADP-1118

## Summary and Scope

This PR makes it possible to call `CoinSelection.performSelection` with an available UTXO set that has some entries pre-selected. The pre-selected entries will be included in the result.

In addition, this PR:
- renames `SelectionInsufficientError` to `SelectionLimitReachedError` to clarify that this error condition relates to the selection limit.

## Limitations

This PR only adjusts the API provided by `CoinSelection` (and `CoinSelection.Balance`). It does not adjust `Wallet.selectAssets` to accept a `UTxOSelection`. This is left to future PRs.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -199,13 +243,98 @@ data SelectionError
+            Just . (view #coin existingInputValue <>) . fromMaybe mempty
+
+        modifyUTxOAvailable :: UTxOIndex -> UTxOIndex
+        modifyUTxOAvailable = UTxOIndex.deleteMany $ UTxO.dom existingInputs
```
Oh, didn't think about that, an existing input means that we've effectively lost that input from our UTxO set.

jonathanknowles replied:
> Oh, didn't think about that, an existing input means that we've effectively lost that input from our UTxO set.

Yes that's right! We definitely shouldn't reselect it.

By removing it here, it means that we don't have to modify the internal algorithm: the internal algorithm can assume that everything in `availableUTxO` really is available for selection.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:
```diff
@@ -199,13 +243,98 @@ data SelectionError
+            modifyUTxOAvailable
+      where
+        modifyAssetsToMint :: TokenMap -> TokenMap
+        modifyAssetsToMint = TokenMap.add (view #tokens existingInputValue)
```
Why are the existing input values added to assets to mint?

jonathanknowles replied:
Good question.

When we have pre-existing inputs, these are effectively "supplying extra value" to a coin selection.

The `assetsToMint` field currently provides the only well-tested way of supplying extra value.

By adding to this field here, we're effectively getting the ability to "add the value from several extra inputs" for free. We can also take advantage of the fact that the `assetsToMint` and `assetsToBurn` fields are already well covered with property tests. In a sense, we're treating the inner algorithm as a black box that we trust will do the right thing.

Of course, it would be nicer if the `assetsToMint` and `assetsToBurn` field names could reflect their generality. So we have a ticket for precisely that: ADP-1117

jonathanknowles committed:
Change `utxoAvailable` to type `UTxOSelection` in `SelectionParams`.

This commit allows callers of `CoinSelection.performSelection` to
pre-select a certain subset of inputs from the available UTxO set.

This pre-selection will form the basis of the input selection
created by the internal `Balance.performSelection` function.

While existing tests all still pass, further work will be required to
verify that `performSelection` behaves as expected if the set of
pre-selected UTxOs is non-empty.

jonathanknowles committed:
Add `available{Balance,UTxO}` accessor functions to `UTxOSelection`.

The "available balance" is simply the sum of the selected and leftover
balances. It predicts what the selected balance would be if every single
UTxO were selected.

Similarly, the "available UTxO set" is the union of the selected and
leftover UTxO sets. It predicts what the selected UTxO set would be if
every single UTxO were selected.

The available balance and the available UTxO set both remain constant
over applications of `select` and `selectMany`.

This commit adds tests to verify that this is the case.

jonathanknowles committed:
Add `genTxInFunction` to `Types.Tx.Gen`.

This commit provides a slightly more intuitive API for generating
functions from `TxIn` to any type 'a'.

Callers no longer need to remember that they must call `genFunction`
with `coarbitraryTxIn`.

jonathanknowles committed:
Use `genTxInFunction` in `Types.UTxOSelection.Gen`.

This commit uses `genTxInFunction` to simplify `genUTxOSelection`.

jonathanknowles committed:
Add function `selectionLimitExceeded`.

This function indicates whether a given selection limit has been
exceeded.

It will be used by further commits.

jonathanknowles committed:
Generate `SelectionParams` with some inputs already selected.

This enables us to test `performSelection` in the case where some inputs
have been pre-selected.

This commit also:

1.  adjusts the post-condition check within `performSelection` to cover the
    case where `performSelection` is called with a selection that is already
    larger than the selection limit.

2.  adds some coverage checks to `prop_performSelection` to ensure that we're
    testing with a variety of pre-selected input conditions.

jonathanknowles committed:
Rename `SelectionInsufficientError` to `SelectionLimitReachedError`.

This error relates to the selection limit, and occurs when:

 - the existing selection size has reached (or exceeded) the selection limit.
 - the existing selection is insufficient to cover the required amount.

The existing name did not represent this concept very clearly, so this
commit renames the error to match the specific condition it represents.

jonathanknowles committed:
Remove now-redundant post-condition check in `performSelection`.

This check is now performed later on, in the case that `runSelection`
returns a selection, so we can remove it from the `Nothing` case.

jonathanknowles committed:
Adjust `prop_runSelection_UTxO_*` properties to work with `UTxOSelection`.

This commit extends the coverage of the `prop_runSelection_UTxO_*`
properties to cover the case where some UTxOs are already selected.

jonathanknowles committed:
Use `UTxOSelection.isSubSelectionOf` in `prop_runSelection_*`.

This commit adjusts the `prop_runSelection_*` properties to assert that
the available UTxO (at the start) is always a subselection of the result.

jonathanknowles committed:
Update expected error condition in integration tests.

When we invoke coin selection with an available UTxO set that is
completely empty, it now always returns the `EmptyUTxO` error.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -1112,7 +1111,7 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
 
         expectResponseCode HTTP.status403 r
-        expectErrorMessage errMsg403TxTooBig r
+        expectErrorMessage errMsg403EmptyUTxO r
```
This doesn't seem right... `fixtureWallet` has UTxOs cc: @jonathanknowles.
